modules/transactionpool/transactionpool.go

```diff
@@ -58,19 +63,17 @@ type (
 )
 
 // New creates a transaction pool that is ready to receive transactions.
-func New(cs modules.ConsensusSet, g modules.Gateway) (tp *TransactionPool, err error) {
+func New(cs modules.ConsensusSet, g modules.Gateway) (*TransactionPool, error) {
 	// Check that the input modules are non-nil.
 	if cs == nil {
-		err = errors.New("transaction pool cannot use a nil state")
-		return
+		return nil, errNilCS
 	}
 	if g == nil {
-		err = errors.New("transaction pool cannot use a nil gateway")
-		return
+		return nil, errNilGateway
```

Comment:
that's what I'm saying -- this is a wrong place. It's a waste of effort. Outside of testing, the `New` functions are called in exactly one place, where it is easy to verify that their inputs are non-nil. These checks are superfluous and force us to write useless tests for the sake of coverage.
If we do somehow pass `nil` to `New`, the result will almost always be an immediate panic. The risk of such a bug making it into a release is minute, and would be discovered quickly by users anyway.
I would give some examples of where nil arguments are far more likely, but I'm afraid you would want to add checks for all of those too.